firmware/library/L1_Drivers/dac.hpp

```diff
@@ -1,19 +1,76 @@
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/pin_configure.hpp"
 
 class DacInterface
 {
  public:
-    virtual bool Initialize();
-    virtual bool WriteDac(uint16_t dac_output);
-    virtual bool SetVoltage(float voltage);
-    // Sine Demonstration Function
-    virtual void SineDac(uint32_t frequency_hz);
-    // Sawtooth Demonstration Function
-    virtual void SawtoothDac(uint32_t frequency_hz);
-    // Triangle Demonstration Function
-    virtual void TriangleDac(uint32_t frequency_hz);
-    virtual void SetBias(bool bias_level);
+    enum BiasLevel: uint8_t
+    {
+        kBiasHigh  = 0,
+        kBiasLow   = 1
+    };
+    virtual void Initialize(void)                   = 0;
+    virtual bool WriteDac(uint16_t dac_output)      = 0;
+    virtual bool SetVoltage(float voltage)          = 0;
+    virtual void SetBias(BiasLevel bias_level)      = 0;
+};
+class Dac : public DacInterface
+{
+ public:
+    static LPC_DAC_TypeDef * dac_ptr;
+    static LPC_SC_TypeDef * sc_ptr;
+    static LPC_IOCON_TypeDef * iocon_ptr;
+    static constexpr float kVref = 3.3;
+    enum Bit : uint8_t
+    {
+        kClockDividerReg = 0,
+        kBiasReg = 16,
+        kDacOutReg = 6,
+        kEnableDacReg = 16
+    };
+    // PinConfigure led3(PinConfigure::CreatePinConfigure<1, 15>());
+    /* Initialize Dac by setting the clock divider and enabling 
+      The Dac Pin withiin IOCON.  Then set the initial Bias level to 0 */
+    void Initialize(void) override
+    {
+        // Set Clock Divider to 1
+        sc_ptr->PCLKSEL = (1 << kClockDividerReg);
+        // Enable DAC Pin P0.26
+        iocon_ptr->P0_26 = (1 << kEnableDacReg);
+        // Set Update Rate to 1MHz
+        dac_ptr->CR &= ~(1 << kBiasReg);
+    }
+    // The Dac ouptut is a 10 bit input and thus it is necessary to
+    // ensure dac_output is less than 1024(largest 10-bit number)
+    bool WriteDac(uint16_t dac_output) override
+    {
+        SJ2_ASSERT_FATAL(dac_output > 1023,
+                         "dac_output is above SJ2"
+                         "output capacity.");
+        dac_ptr->CR = (dac_output << kDacOutReg);
+        return true;
+    }
+    // Takes an input voltage and converts the float value and calculates
+    // the conversion necessary and then typecasts it to an integer.
+    // If the voltage value is greater than 3.3 it will fail and end.
+    bool SetVoltage(float voltage) override
+    {
+        // value = (dac_out * 1024)/VrefP
+        float value = (voltage * 1024) / kVref;
+        int   conversion = static_cast<int>(value);
+        SJ2_ASSERT_FATAL(voltage > kVref,
+                         "Input voltage above SJ2"
+                         "output capacity of 3.3 Volts.");
+        dac_ptr->CR = (conversion << kDacOutReg);
+        return true;
+    }
+    // Sets the Bias for the Dac, which determines the settling time
+    // , max current, and the allowed maximum update rate
+    void SetBias(BiasLevel bias_level) override
```

Comment:
If you don't make BiasLevel an `enum class`, the user can put in any integer value they like for this, and since you are not using an AND to remove any of the other bits, this can result in additional bits being manipulated. Its better to add that `class` keyword to prevent the user from getting to that point.